### DISCUSSION TITLE
Authentication inside root Layout

### DISCUSSION THREAD

**Discussion by Akumzy** [vercel/next.js#50507] [App Router]
I have been steering at this example on the doc but I'm yet to figure ` getUser()` was implemented leaving me with questions
- Is `getUser` making an API request?
- Is there any function the get user session?

<img width="676" alt="image" src="https://github.com/vercel/next.js/assets/25849232/e772463c-6fb7-449d-83f3-0c2d5de21698">

I really want to understand more, please

**Comment by devjiwonchoi** [comment]
The `getUser` function on the example is just an example of a custom Authentication.
The example code is presenting that in your `layout.tsx`, you can use "your own method of Authentication" 
in order to check if the user is logged in or not, and conditionally return whether to show the `dashboard` or the `login` page.

For your first question "Is getUser making an API request?"
The answer depends on the auth verification method you choose, whether to store login info on the client-side cookie or on the backend server session, and more.

Answer to your second question "Is there any function to get a user session?"
You might want to look up [OAuth](https://oauth.net/) since it offers you overall every necessary validation and authentication method you need for production.

**Reply by Akumzy** [reply]
`OAuth` use Context API and I don't know how that's possible or can be used inside `Layout` assuming it's inside the page or old `_app.js` file that'll be easy for me to understand

**Reply by devjiwonchoi** [reply]
You are right. Therefore I use [NextAuth](https://next-auth.js.org/) for the issue you've described.

**Reply by Akumzy** [reply]
Still, with `NextAuth` to the best of my knowledge, you still need Context API or pass down the `request` object to it which there's still no way to do in Layout except if you're using their [NextAuth ](https://next-auth.js.org/configuration/nextjs)

**Reply by devjiwonchoi** [reply]
Returning to your primary question, `getUser` function is not a Next.js feature but just an example of Authentication, 
and therefore you may choose any Authentication method you'd like to use.

For authentication inside the root layout, I created `AuthProvider` and wrapped it around the main components.

Example:
```
// [app/layout.tsx]
// ...
export default function RootLayout({children}) {
  return(
    <html>
      <body>
        <AuthProvider>
          <Header />
          <main>{children}</main>
          <Footer />
       </AuthProvider>
// ...
}
```

ref: https://www.youtube.com/watch?v=Eh3EpwqT4cM&ab_channel=HamedBahram
This video helped me a lot, and hope it might resolve your question.

**Reply by Akumzy** [reply]
First thank you so much for replaying back to me back to back
But I want to utilize the pattern showcased in the screenshot I shared
Anyway I guess I'll use the `cookies` function
```ts
  const cookieStore = cookies();
  const hasAuth = !!cookieStore.get('auth')?.value;
```
This seemed to work, hopefully, there will not be any gotcha while using it
